### PR TITLE
Mask tokens when exporting onyx state

### PR DESCRIPTION
**Pull request by TMisiukiewicz**
### Details

For security reasons, `authToken` and `encryptedAuthToken` should be always masked, regardless of whether the "Mask fragile user data" option is enabled.

This is a follow up to a discussion here: https://expensify.slack.com/archives/C05LX9D6E07/p1724247139893599
### Fixed Issues

$ https://github.com/Expensify/App/issues/47995
PROPOSAL:

### Tests

1. Go to Troubleshoot section
2. Make sure "Mask fragile user data while exporting Onyx state" option is off
3. Export Onyx state
4. Open the file and find `authToken` and `encryptedAuthToken`
5. Verify both properties have `***` value
- [x] Verify that no errors appear in the JS console

### QA Steps

1. Go to Troubleshoot section
2. Make sure "Mask fragile user data while exporting Onyx state" option is off
3. Export Onyx state
4. Open the file and find `authToken` and `encryptedAuthToken`
5. Verify both properties have `***` value
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/639a1b69-c0c2-4bca-8094-2792fae74658

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/6a1bcf9a-d47e-4be9-96bd-2282724b5b0a

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/5e7acd13-1f60-453d-8859-22e8815c62fc

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/1384fb2a-1eb1-4d80-aad4-e22b7febfe9b

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/cc612833-53e8-4204-8789-26dabad1d3f2

</details>

### CODE REVIEW

**Comment by mountiny** [comment]
Asked @TMisiukiewicz to add a unit test so we have another layer of protection in case anyone would remove this

**Comment by TMisiukiewicz** [comment]
@mountiny @hoangzinh done ✅

File: src/libs/ExportOnyxState/common.ts [hunk 1]
```diff
@@ -1,5 +1,22 @@
...
+import type {Session} from '@src/types/onyx';
+
+const maskSessionDetails = (data: Record<string, unknown>): Record<string, unknown> => {
+    const session = data.session as Session;
+
+    Object.keys(session).forEach((key) => {
+        if (key !== 'authToken' && key !== 'encryptedAuthToken') {
+            return;
+        }
+        session[key] = '***';
```
Comment (adhorodyski) [inline src/libs/ExportOnyxState/common.ts:12]: what about extracting this pattern to a global constant?

**Review by adhorodyski — APPROVED** [review]

File: src/libs/ExportOnyxState/common.ts [hunk 2]
```diff
@@ -1,5 +1,24 @@
 import {Str} from 'expensify-common';
 import ONYXKEYS from '@src/ONYXKEYS';
+import type {Session} from '@src/types/onyx';
+
+const MASKING_PATTERN = '***';
+
+const maskSessionDetails = (data: Record<string, unknown>): Record<string, unknown> => {
+    const session = data.session as Session;
```
[the remark this answers is not in the document]
Comment (TMisiukiewicz) [reply src/libs/ExportOnyxState/common.ts:8]: updated to not modify original data, thanks 👍

File: src/libs/ExportOnyxState/common.ts [hunk 3]
```diff
@@ -29,6 +50,16 @@ const maskFragileData = (data: Record<string, unknown>, parentKey?: string): Rec
...
+    onyxState = maskSessionDetails(onyxState);
+    // Mask fragile data other than session details if the user has enabled the option
+    if (isMaskingFragileDataEnabled) {
+        onyxState = maskFragileData(onyxState);
+    }
+
+    return onyxState;
 };
+
+export default maskOnyxState;
```
[the remark this answers is not in the document]
Comment (TMisiukiewicz) [reply src/libs/ExportOnyxState/common.ts]: good idea, done 👍

**Comment by TMisiukiewicz** [comment]
> I think It's intentional. If it happens in main branch as well, I think we should leave it as it is.

Hmm I couldn't find any reasoning about it, I think it's was an oversight. Look at the keys from exported Onyx state on mobile:
image

Each key is a number, makes it superhard to extract any Onyx key in an efficient way. On web Onyx keys are assigned correctly:
image

Having in mind we will work on implementing the "import state" feature https://github.com/Expensify/App/issues/47786, I think we should align the export across all platforms to make it work properly regardless of the platform the state was exported from.

> Btw, I watched your recording for ios/android. I found that it was exported to a text format. But in my case, it's "shm" & "wall" files

I think those files are automatically generated when you make a dump of the database, however my videos are recorded with a fix mentioned above. In your case it creates a dump, but it cannot find the `session` key because of the invalid Onyx state structure, so the txt is not created

**Comment by mountiny** [comment]
> @hoangzinh seems like a difference between exported file on native platforms and web, the Onyx State file is built in a different shape than on the web/desktop. Do you want to align it while working on this PR or should we address it in a separate PR? From my perspective it's safe to fix this here, as currently on main branch the state is exported in a readable format, it's just different than on the web, so it's just a matter of unifying it.

@TMisiukiewicz lets unify it here, no rush on getting the current change out without unified format

**Review by hoangzinh — APPROVED** [review]
LGTM

File: src/libs/ExportOnyxState/common.ts [hunk 4]
```diff
@@ -29,6 +50,16 @@ const maskFragileData = (data: Record<string, unknown>, parentKey?: string): Rec
     return maskedData;
 };
 
-export default {
-    maskFragileData,
+const maskOnyxState = (data: Record<string, unknown>, isMaskingFragileDataEnabled?: boolean) => {
+    let onyxState = data;
+    // Mask session details by default
```
Comment (mountiny) [inline src/libs/ExportOnyxState/common.ts:55]:
```suggestion

    // Mask session details by default
```

File: src/libs/ExportOnyxState/common.ts [hunk 5]
```diff
@@ -29,6 +50,16 @@ const maskFragileData = (data: Record<string, unknown>, parentKey?: string): Rec
...
+    onyxState = maskSessionDetails(onyxState);
+    // Mask fragile data other than session details if the user has enabled the option
```
Comment (mountiny) [inline src/libs/ExportOnyxState/common.ts:57]:
```suggestion

    // Mask fragile data other than session details if the user has enabled the option
```

**Review by mountiny — APPROVED** [review]
Just small NABs that we can skip for now but lets try to leave space above the comments for better readability.

Thanks!

**Pull request merged**